### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by mirceanis**
Implement the ENS resolution snap.

The snap does forward and reverse resolution using the ENS protocol.
ENS is rooted in Ethereum mainnet, so resolution should always[^1] start by querying the mainnet contracts.
To that end, this snap instantiates its own `ethers#InfuraProvider` when the currently selected `chainId` is not mainnet.

This behavior will be removed when there is a multi-chain provider API available for snaps, which should also improve privacy, as the snap will end up using the user-configured connections (if any).

This snap is meant to be preinstalled in the MM extension and mobile apps.

Initial e2e tests use live networks.

---

This should be reviewed after #2 is merged, to simplify the review process.

[^1]: or Sepolia or Holesky for testing

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Lines starting with '#' are comments.
+# Each line is a file pattern followed by one or more owners.
+
+*                                    mirceanis
```
Comment (Mrtenz) [inline .github/CODEOWNERS:4]: This seems to be invalid:

> Invalid owner on line 4

Ideally this should be a team within the MetaMask org.

File: packages/snap/snap.manifest.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/MetaMask/ens-resolver-snap.git"
+  },
+  "source": {
+    "shasum": "CV59Ej5jBOSHvRT1EU3MA9rDVgK3x6FoB5W3H8r9HHY=",
+    "location": {
+      "npm": {
+        "filePath": "dist/bundle.js",
+        "packageName": "ens-resolver-snap",
```
Comment (Mrtenz) [inline packages/snap/snap.manifest.json:14]:
```suggestion
        "packageName": "@metamask/ens-resolver-snap",
```

File: packages/snap/package.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
+{
+  "name": "ens-resolver-snap",
```
Comment (Mrtenz) [inline packages/snap/package.json:2]:
```suggestion
  "name": "@metamask/ens-resolver-snap",
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: packages/snap/package.json (hunk 3 above)
Comment (mirceanis) [reply packages/snap/package.json]: done

File: packages/snap/snap.manifest.json (hunk 2 above)
Comment (mirceanis) [reply packages/snap/snap.manifest.json]: done

File: .github/CODEOWNERS (hunk 1 above)
Comment (mirceanis) [reply .github/CODEOWNERS]: done. I set `@MetaMask/engineering` for now

**Review by mirceanis — COMMENTED** [review]
I addressed all comments

**Review by Mrtenz — DISMISSED** [review]

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**